### DISCUSSION TITLE
Modals always need a close button? What constitutes a close button?

### DISCUSSION THREAD

**Discussion by jcklpe** [uswds/uswds#4570] [Q&A]
The USWDS Modal guidelines say:

"there needs to be a clear way to accept or reject the content in the modal"

I have a two step modal that looks sorta like this: 
![image](https://user-images.githubusercontent.com/9451726/157344493-0d1107b1-bfc6-4d03-895b-d892d17db976.png)

A teammate has raised a concern that this does not follow the above quoted guidelines because the X button isn't explicit enough. All the examples given on the USWDS guideline pages are giving binary choices (login/logout, save/don't save) 

![image](https://user-images.githubusercontent.com/9451726/157344902-402e1a69-8a27-4697-b585-950eab6a42ed.png)

They feel that the close button should also be an explicit choice like this:
![image](https://user-images.githubusercontent.com/9451726/157345125-713075fb-6c03-4fc4-b7d3-58f19f603a46.png)

I'm not asking ya'll to sit here King Solomon style and adjudicate this specific case for us, but I was wondering if ya'll could provide some clarifications (and in turn this clarification might also be a helpful addition to the docs). 

Is the guideline either:

- There must be at least one way to back out of a modal (and this can include an X button) 
- There must be at least one way to back out of a modal (and the X button does not count)

**Comment by brunerae** [comment]
Hey @jcklpe ! Great question. Let me raise this with the team and see if they have any thoughts.

**Comment by mejiaj** [comment]
@jcklpe not a UX expert, so happy to be corrected but there are several ways to get out of the modal. You can close via <kbd>escape</kbd>, by clicking on overlay, and the close button.

The close button has an aria-label that says _Close this window_, I don't know if adding another with similar text would help.

**Reply by jcklpe** [reply]
Thanks @mejiaj : can you help clarify if having a close button with an aria label is sufficient to follow the design system or if all modals are required to have an explicit textual exit button in addition to the "X" button?

**Reply by mejiaj** [reply]
Your approach makes sense. I would just modify the language to something like _Close without saving_ or something explaining what the close action will do.

Saying just _Close_ in a multi-step process doesn't seem very clear.

**Comment by ryparker** [comment]
An accessible X button can provide the way to dismiss a modal. A second text button is not universally required: the [ARIA modal-dialog guidance](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/) recommends a visible button in the tab sequence that closes the dialog, which can be a close icon or a Cancel button.

The X needs to be a real button with a meaningful accessible name. Users also need the expected keyboard dismissal and focus behavior. An icon's shape alone does not provide its accessible name.

For the two-step workflow, the other question is whether the consequence is clear. If closing discards progress, visible text such as “Cancel” or “Discard changes” may communicate that better than an X alone. An accessible name such as “Close without saving” helps screen-reader users, but does not make that consequence visible to everyone.

So the close icon can count as dismissal. Whether it communicates this workflow well is a separate usability decision.
